File: src-native/src/core/Engine.cpp
#include "yuga/core/Engine.hpp"
#include "yuga/core/Window.hpp"
#include "yuga/graphics/Renderer.hpp"
#include "yuga/scene/SceneManager.hpp"
#include "yuga/resource/ResourceManager.hpp"
#include "yuga/input/InputManager.hpp"
#include "yuga/audio/AudioManager.hpp"
#include "yuga/physics/PhysicsSystem.hpp"
#include "yuga/ecs/EntityComponentSystem.hpp"

#include <chrono>

namespace yuga {
namespace core {

Engine::Engine()
    : m_isRunning(false)
    , m_timeStep(1.0f / 60.0f)  // 60 FPS fixed timestep
    , m_accumulatedTime(0.0f)
{
}

Engine::~Engine() {
    shutdown();
}

bool Engine::initialize(const std::string& appName, int width, int height) {
    // Initialize in reverse order of dependencies
    
    // 1. Window and Input (platform layer)
    m_window = std::make_unique<Window>();
    if (!m_window->initialize(appName, width, height)) {
        return false;
    }

    m_inputManager = std::make_unique<InputManager>(m_window.get());
    if (!m_inputManager->initialize()) {
        return false;
    }

    // 2. Graphics system
    m_renderer = std::make_unique<Renderer>();
    if (!m_renderer->initialize(m_window.get())) {
        return false;
    }

    // 3. Resource management
    m_resourceManager = std::make_unique<ResourceManager>();
    if (!m_resourceManager->initialize()) {
        return false;
    }

    // 4. Core systems
    m_audioManager = std::make_unique<AudioManager>();
    if (!m_audioManager->initialize()) {
        return false;
    }

    m_physicsSystem = std::make_unique<PhysicsSystem>();
    if (!m_physicsSystem->initialize()) {
        return false;
    }

    // 5. Entity-Component System
    m_ecs = std::make_unique<EntityComponentSystem>();
    if (!m_ecs->initialize()) {
        return false;
    }

    // 6. Scene Management (depends on ECS)
    m_sceneManager = std::make_unique<SceneManager>(this);
    if (!m_sceneManager->initialize()) {
        return false;
    }

    m_isRunning = true;
    return true;
}

void Engine::run() {
    using clock = std::chrono::high_resolution_clock;
    using duration = std::chrono::duration<float>;

    auto lastTime = clock::now();

    while (m_isRunning) {
        auto currentTime = clock::now();
        float frameTime = std::chrono::duration_cast<duration>(currentTime - lastTime).count();
        lastTime = currentTime;

        // Cap maximum frame time to avoid spiral of death
        if (frameTime > 0.25f) {
            frameTime = 0.25f;
        }

        m_accumulatedTime += frameTime;

        // Fixed timestep update
        while (m_accumulatedTime >= m_timeStep) {
            update(m_timeStep);
            m_accumulatedTime -= m_timeStep;
        }

        // Render at display refresh rate
        render();

        // Process window events and check if we should continue
        m_isRunning = processFrame();
    }
}

void Engine::shutdown() {
    // Shutdown in reverse order of initialization
    if (m_sceneManager) m_sceneManager->shutdown();
    if (m_ecs) m_ecs->shutdown();
    if (m_physicsSystem) m_physicsSystem->shutdown();
    if (m_audioManager) m_audioManager->shutdown();
    if (m_resourceManager) m_resourceManager->shutdown();
    if (m_renderer) m_renderer->shutdown();
    if (m_inputManager) m_inputManager->shutdown();
    if (m_window) m_window->shutdown();
}

bool Engine::processFrame() {
    m_inputManager->processEvents();
    return !m_window->shouldClose();
}

void Engine::update(float deltaTime) {
    // Update subsystems in dependency order
    m_inputManager->update(deltaTime);
    m_physicsSystem->update(deltaTime);
    m_ecs->update(deltaTime);
    m_sceneManager->update(deltaTime);
    m_audioManager->update(deltaTime);
}

void Engine::render() {
    m_renderer->beginFrame();
    m_sceneManager->render();
    m_renderer->endFrame();
}

} // namespace core
} // namespace yuga

File: src-native/src/core/engine.rs
use crate::core::{
    subsystem::SubsystemManager,
    ecs::{EcsSystem, Transform, RigidBody, Renderable},
    scripting::ScriptingSystem,
};
use crate::graphics::Renderer;
use std::time::Instant;

pub struct Engine {
    subsystem_manager: SubsystemManager,
    running: bool,
}

impl Engine {
    pub fn new() -> Result<Self, String> {
        let mut engine = Self {
            subsystem_manager: SubsystemManager::new(),
            running: false,
        };

        // Initialize subsystems
        engine.init_subsystems()?;

        Ok(engine)
    }

    fn init_subsystems(&mut self) -> Result<(), String> {
        // Initialize subsystems
        let mut ecs = EcsSystem::new();
        let mut physics = PhysicsSystem::new();
        let mut audio = AudioSystem::new().map_err(|e| e.to_string())?;
        let mut input = InputSystem::new();
        let mut assets = AssetSystem::new("assets/");
        let mut profiling = ProfilingSystem::new();

        // Set up example entity
        let entity = ecs.spawn((
            Transform {
                position: nalgebra::Vector3::new(0.0, 0.0, 0.0),
                rotation: nalgebra::UnitQuaternion::identity(),
                scale: nalgebra::Vector3::new(1.0, 1.0, 1.0),
            },
            RigidBody {
                velocity: nalgebra::Vector3::zeros(),
                acceleration: nalgebra::Vector3::zeros(),
                mass: 1.0,
            },
            Renderable {
                mesh_id: 0,
                material_id: 0,
            },
        ));

        // Register subsystems
        self.subsystem_manager.register(ecs)?;
        self.subsystem_manager.register(physics)?;
        self.subsystem_manager.register(audio)?;
        self.subsystem_manager.register(input)?;
        self.subsystem_manager.register(assets)?;
        self.subsystem_manager.register(profiling)?;

        // Initialize scripting system
        let scripting = ScriptingSystem::new();
        self.subsystem_manager.register(scripting)?;

        // Initialize other subsystems as needed

        Ok(())
    }

    pub fn run(&mut self) -> Result<(), String> {
        self.running = true;
        let mut last_time = Instant::now();

        while self.running {
            let current_time = Instant::now();
            let delta_time = (current_time - last_time).as_secs_f32();
            last_time = current_time;

            // Start frame profiling
            if let Some(profiling) = self.subsystem_manager.get_mut::<ProfilingSystem>() {
                profiling.begin_frame();
            }

            // Process input
            if let Some(input) = self.subsystem_manager.get_mut::<InputSystem>() {
                profiling.begin_subsystem_profile("Input");
                if input.is_key_just_pressed(glfw::Key::Escape) {
                    self.running = false;
                }
                profiling.end_subsystem_profile("Input");
            }

            // Update physics
            if let Some(physics) = self.subsystem_manager.get_mut::<PhysicsSystem>() {
                profiling.begin_subsystem_profile("Physics");
                physics.update(delta_time)?;
                profiling.end_subsystem_profile("Physics");
            }

            // Update audio
            if let Some(audio) = self.subsystem_manager.get_mut::<AudioSystem>() {
                profiling.begin_subsystem_profile("Audio");
                audio.update(delta_time)?;
                profiling.end_subsystem_profile("Audio");
            }

            // Update ECS
            if let Some(ecs) = self.subsystem_manager.get_mut::<EcsSystem>() {
                profiling.begin_subsystem_profile("ECS");
                ecs.update(delta_time)?;
                profiling.end_subsystem_profile("ECS");
            }

            // Update asset system
            if let Some(assets) = self.subsystem_manager.get_mut::<AssetSystem>() {
                profiling.begin_subsystem_profile("Assets");
                assets.update(delta_time)?;
                profiling.end_subsystem_profile("Assets");
            }

            // End frame profiling and collect stats
            if let Some(profiling) = self.subsystem_manager.get_mut::<ProfilingSystem>() {
                profiling.end_frame();
                
                // Log performance stats every second
                if current_time.elapsed().as_secs_f32() >= 1.0 {
                    let stats = profiling.get_stats();
                    println!(
                        "FPS: {:.2}, Frame Time: {:.2}ms, Draw Calls: {}, Triangles: {}",
                        stats.fps,
                        stats.frame_time.as_secs_f32() * 1000.0,
                        stats.draw_calls,
                        stats.triangle_count
                    );
                }
            }
        }

        self.shutdown()?;
        Ok(())
    }

    pub fn shutdown(&mut self) -> Result<(), String> {
        self.running = false;
        self.subsystem_manager.shutdown()
    }
}